pallets/asset-tx-payment/src/payment.rs

```diff
@@ -0,0 +1,125 @@
+	transaction_validity::InvalidTransaction,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+pub trait HandleCredit<AccountId, B: Balanced<AccountId>> {
+	fn handle_credit(credit: CreditOf<AccountId, B>);
+}
+
+/// Handle withdrawing, refunding and depositing of transaction fees.
+pub trait OnChargeAssetTransaction<T: Config> {
+	/// The underlying integer type in which fees are calculated.
+	type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+	type AssetId: FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default + Eq;
+	type LiquidityInfo;
+
+	/// Before the transaction is executed the payment of the transaction fees
+	/// need to be secured.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		asset_id: Self::AssetId,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError>;
+
+	/// After the transaction was executed the actual fee can be calculated.
+	/// This function should refund any overpaid fees and optionally deposit
+	/// the corrected amount.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn correct_and_deposit_fee(
+		who: &T::AccountId,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		post_info: &PostDispatchInfoOf<T::Call>,
+		corrected_fee: Self::Balance,
+		tip: Self::Balance,
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError>;
+}
+
+/// Implements the transaction payment for a module implementing the `Currency`
+/// trait (eg. the pallet_balances) using an unbalance handler (implementing
+/// `OnUnbalanced`).
+///
+/// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
+/// then tip.
+pub struct FungiblesAdapter<CON, HC>(PhantomData<(CON, HC)>);
+
+/// Default implementation for a Currency and an OnUnbalanced handler.
+///
+/// The unbalance handler is given 2 unbalanceds in [`OnUnbalanced::on_unbalanceds`]: fee and
+/// then tip.
+impl<T, CON, HC> OnChargeAssetTransaction<T> for FungiblesAdapter<CON, HC>
+where
+	T: Config,
+	CON: BalanceConversion<BalanceOf<T>, AssetIdOf<T>, AssetBalanceOf<T>>,
+	HC: HandleCredit<T::AccountId, T::Fungibles>,
+	AssetIdOf<T>: FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default + Eq,
+{
+	type LiquidityInfo = CreditOf<T::AccountId, T::Fungibles>;
+	type Balance = BalanceOf<T>;
+	type AssetId = AssetIdOf<T>;
+
+	/// Withdraw the predicted fee from the transaction origin.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		_call: &T::Call,
+		_info: &DispatchInfoOf<T::Call>,
+		asset_id: Self::AssetId,
+		fee: Self::Balance,
+		_tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		let converted_fee = CON::to_asset_balance(fee, asset_id)
```

Comment:
I guess this is what will handle the conversion rate between token and a particular asset, right? It's going to be oracle-based (or dex-based?)?

Reply:
Right now it's based on the ratio between minimum balances.
It could be oracle-based in the future (DEX-based seems very unlikely)

Reply:
@tomusdrw @shawntabrizi I (finally!) ran a test via Polkadot-js and ran into a practical limitation: even if the conversion is mathematically correct it can end up rounding down the fee to 0 if the asset has low precision. I wonder whether we should
1. just not allow assets with low precision (i.e. low `min_balance`) to be `is_sufficient`
2. OR add `.max(1)` here

Opinions?

Reply:
Couldn't we reject the transaction (i.e. return it as invalid) in case it returns a `0` fee? I think that's the most sensible and would also protect us in the future in case the `to_asset_balance` method is implemented differently.

Reply:
I would say doing `.max(1)` ist strictly better because the transaction is not invalid and the fee is calculated "correctly" (the appropriate fee would be less than 1 token).

Reply:
Though, yes, we could reject the transaction in principle.